akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -1148,6 +1159,20 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
       }
       internalInvoke(event)
     }
+
+    // external call
+    def cancel(): Unit = {
+      currentState.get() match {
+        case Completed ⇒
+        case previous ⇒
+          if (!currentState.compareAndSet(previous, Completed)) cancel()
+          else {
+            releaseCallback(this)
+            currentState.set(Completed)
```

Comment:
wasn't that just set to Completed by the compareAndSet?

Reply:
Indeed